libass/ass_shaper.c

```diff
@@ -682,9 +691,19 @@ static bool shape_harfbuzz(ASS_Shaper *shaper, GlyphInfo *glyphs, size_t len)
                 level == shaper->emblevels[i + 1])
             i++;
 
+        int lead_context = 0, trail_context = 0;
+        if (offset > 0 && !glyphs[offset].starts_new_run &&
+                is_shaping_control(glyphs[offset - 1].symbol))
+            lead_context = 1;
+        if (i < (len - 1) && !glyphs[i + 1].starts_new_run &&
+                is_shaping_control(glyphs[i + 1].symbol))
+            trail_context = 1;
+
         hb_buffer_pre_allocate(buf, i - offset + 1);
-        hb_buffer_add_utf32(buf, shaper->event_text + offset, i - offset + 1,
-                0, i - offset + 1);
+        hb_buffer_add_utf32(buf,
+                shaper->event_text + offset - lead_context,
+                i - offset + 1 + lead_context + trail_context,
+                lead_context, i - offset + 1);
```

Comment:
Maybe I'm misunderstanding the harfbuzz-API, but shouldn't the last parameter, item length, be `i - offset + 1 - trail_context` to not include the trail in the buffer, but only for context? *(as done for `lead_context`)*

Reply:
`i - offset + 1` is the length of the item. Compare to `i - offset + 1 + lead_context + trail_context` in the line above, which is the length of the item _plus_ context.

The code above peeks at the leading and trailing context without incrementing `i`.

The previous argument, `lead_context`, is the offset from the start of the “text” (the pointer in the second argument) to the start of the item. The whole “text” in this case is the item plus both leading and trailing context.